### ISSUE TITLE
Deoplete throws Python Exception on closing

### ISSUE THREAD

**Issue by TheSovietStorm** [Shougo/deoplete.nvim#643]
# Problems summary
When I close vim, I get a Python exception.
*many probably irrelevant lines*
*myPath*.vim/plugged/deoplete.nvim/rplugin/pytho
n3/deoplete/process.py", line 55, in enqueue_output
[deoplete@yarp]     b = self._proc.stdout.read(1)
[deoplete@yarp] AttributeError: 'NoneType' object has no attribute 'stdout'
*other hopefully irrelevant lines*
## Expected
Close vim and don't get any error message.

## Environment Information

 * deoplete version(SHA1): [3dee4b1](https://github.com/Shougo/deoplete.nvim/commit/3dee4b1fd272e07aab945ac514625080567e9019)
 * OS: macOS

 * neovim/Vim version:Vim 8.0.1450

 * `:checkhealth` or `:CheckHealth` result(neovim only):

## Provide a minimal init.vim/vimrc with less than 50 lines (Required!)

```vim
" Your minimal init.vim/vimrc
set runtimepath+=~/path/to/deoplete.nvim/
let g:deoplete#enable_at_startup = 1
let g:deoplete#sources#clang#libclang_path = '/Library/Developer/CommandLineTools/usr/lib/libclang.dylib'
let g:deoplete#sources#clang#clang_header = '/Applications/Xcode.app/Contents/Developer/Platforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include'

let g:deoplete#sources#rust#racer_binary='/Users/*me*/.cargo/bin/racer'
let g:deoplete#sources#rust#rust_source_path='/Users/*me*/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src'
```

## The reproduce ways from neovim/Vim starting (Required!)

1. Open vim
2. Close vim
3. Look at terminal
3a. (Don't do this in Tmux or Fullscreen, as the error is only visible in windowed mode for me

## Generate a logfile if appropriate

It doesn't show up for the provided steps, so I had to copy paste the relevant lines in the problem description

## Screen shot (if possible)
<img width="569" alt="bildschirmfoto 2018-02-07 um 22 18 24" src="https://user-images.githubusercontent.com/20505981/35942139-f567d3fc-0c54-11e8-9185-b2e880f87478.png">

**Comment by Shougo** [comment]
Please test the latest version.

**Comment by TheSovietStorm** [comment]
Thanks it works now 👍

**Comment by dennisbakhuis** [comment]
Hi Shougo,
I have this same issue as described in this post. Everything works fine at runtime, but when I close Vim, I get a huge error/warning. What I noticed is that I can only see the error because it does not fit in a single window and 'more' is used to scroll through the error. When the window is large, vim restores the shell screen to before vim started.

My system: OSX 10.13.3, vim 8.0.1500, latest Deoplete.nvim installed using Vim-Plug.
Screenshot of the error screen:
https://www.dropbox.com/s/pcxex2nd5yyx3s4/Screen%20Shot%202018-02-12%20at%2012.51.01.png?dl=0

Thanks for the great work!

**Comment by TheSovietStorm** [comment]
@dennisbakhuis I would guess the screenshot is wrong 😃

**Comment by TheSovietStorm** [comment]
Oh and by the way, after updating to the latest deoplete version I get the following error on instantly closing vim after opening it:

deoplete@yarp] error caught in notification handler 'deoplete_on_event [{'max_menu_w
idth': 56, 'cwd': '/Users/myName', 'vars': {'deoplete#_serveraddr': '127.0.0.1:
50777', 'deoplete#_prev_completion': {'complete_position': [], 'candidates': [], 'eve
nt': ''}, 'deoplete#_initialized': 1, 'deoplete#auto_complete_delay': 50, 'deoplete#m
ax_abbr_width': 80, 'deoplete#enable_yarp': 0, 'deoplete#omni#input_patterns': {}, 'd
eoplete#sources#rust#racer_binary': '/Users/myName/.cargo/bin/racer', 'deoplete
#auto_refresh_delay': 50, 'deoplete#disable_auto_complete': 0, 'deoplete#omni_pattern
s': {}, 'deoplete#enable_refresh_always': 0, 'deoplete#_keyword_patterns': {'_': '[a-
zA-Z_]\\k*'}, 'deoplete#_logging': {}, 'deoplete#auto_complete_start_length': 2, 'deo
plete#sources#rust#documentation_max_height': 20, 'deoplete#_channel_id': 1, 'deoplet
e#enable_smart_case': 0, 'deoplete#enable_at_startup': 1, 'deoplete#max_menu_width':
40, 'deoplete#enable_ignore_case': 0, 'deoplete#complete_method': 'complete', 'deople
te#omni#functions': {}, 'deoplete#_rank': {}, 'deoplete#enable_on_insert_enter': 1, '
deoplete#sources#clang#clang_header': '/Applications/Xcode.app/Contents/Developer/Pla
tforms/MacOSX.platform/Developer/SDKs/MacOSX.sdk/usr/include', 'deoplete#ignore_sourc
es': {}, 'deoplete#max_list': 100, 'deoplete#member#prefix_patterns': {}, 'deoplete#d
elimiters': ['/'], 'deoplete#keyword_patterns': {}, 'deoplete#enable_camel_case': 0,
'deoplete#num_processes': 4, 'deoplete#sources#rust#rust_source_path': '/Users/myName/.rustup/toolchains/nightly-x86_64-apple-darwin/lib/rustlib/src/rust/src', 'deo
plete#sources': {}, 'deoplete#enable_profile': 0, 'deoplete#_context': {}, 'deoplete#
sources#clang#libclang_path': '/Library/Developer/CommandLineTools/usr/lib/libclang.d
ylib', 'deoplete#_omni_patterns': {'xhtml': ['<', '</', '<[^>]*\\s[[:alnum:]-]*'], 'x
ml': ['<', '</', '<[^>]*\\s[[:alnum:]-]*'], 'html': ['<', '</', '<[^>]*\\s[[:alnum:]-
]*']}, 'deoplete#skip_chars': [], 'deoplete#sources#rust#show_duplicates': 1}, 'same_
filetypes': [], 'smartcase': 0, 'runtimepath': '/Users/myName/.vim,/Users/myName/.vim/plugged/vim-sensible/,/Users/myName/.vim/plugged/vim-fugitive/,/Us
ers/myName/.vim/plugged/nerdtree/,/Users/myName/.vim/plugged/vim-easymoti
on/,/Users/myName/.vim/plugged/vim-airline/,/Users/myName/.vim/plugged/vi
m-airline-themes/,/Users/myName/.vim/plugged/vim-devicons/,/Users/myName/
.vim/plugged/vim-colorschemes/,/Users/myName/.vim/plugged/dash.vim/,/Users/pete
rzdankin/.vim/plugged/vim-maktaba/,/Users/myName/.vim/plugged/vim-bazel/,/Users
/myName/.vim/plugged/vim-misc/,/Users/myName/.vim/plugged/deoplete.nvim/,
/Users/myName/.vim/plugged/nvim-yarp/,/Users/myName/.vim/plugged/vim-hug-
neovim-rpc/,/Users/myName/.vim/plugged/deoplete-clang/,/Users/myName/.vim
/plugged/deoplete-rust/,/Users/myName/.vim/plugged/tagbar/,/Users/myName/
.vim/plugged/gruvbox/,/Users/myName/.vim/plugged/ctrlp.vim/,/usr/local/share/vi
m/vimfiles,/usr/local/share/vim/vim80,/usr/local/Cellar/vim/8.0.1450_1/share/vim/vim8
0/pack/dist/opt/matchit,/usr/local/share/vim/vimfiles/after,/Users/myName/.vim/
after', 'dp_main': '/Users/myName/.vim/plugged/deoplete.nvim/rplugin/python3/de
oplete/dp_main.py', 'bufnr': 1, 'changedtick': -1, 'custom': {'source': {'_': {}}}, '
omni__omnifunc': '', 'filetype': 'nothing', 'input': '', 'is_windows': False, 'delay'
: 50, 'bufsize': 0, 'max_kind_width': 56, 'rpc': 'deoplete_on_event', 'bufpath': '',
'bufname': '', 'bufvars': {}, 'event': 'VimLeavePre', 'max_abbr_width': 56, 'sources'
: [], 'complete_str': '', 'camelcase': 0, 'ignorecase': 0, 'filetypes': ['nothing'],
'keyword_patterns': '[a-zA-Z_][a-zA-Z@0-9_À-ÿ]*', 'dict__dictionary': '', 'position':
 [0, 1, 1, 0], 'next_input': '', 'encoding': 'utf-8'}]'
[deoplete@yarp] Traceback (most recent call last):
[deoplete@yarp]   File "/Users/myName/.vim/plugged/nvim-yarp/pythonx/yarp.py",
line 45, in on_notification
[deoplete@yarp]     getattr(module_obj, method)(*args)
[deoplete@yarp]   File "/Users/myName/.vim/plugged/deoplete.nvim/rplugin/python
3/deoplete/__init__.py", line 64, in deoplete_on_event
[deoplete@yarp]     global_deoplete.on_event(context)
[deoplete@yarp]   File "/Users/myName/.vim/plugged/deoplete.nvim/rplugin/python
3/deoplete/deoplete.py", line 206, in on_event
[deoplete@yarp]     parent.on_event(context)
[deoplete@yarp]   File "/Users/myName/.vim/plugged/deoplete.nvim/rplugin/python
3/deoplete/parent.py", line 67, in on_event

**Comment by blueyed** [comment]
@TheSovietStorm 
Looks like your Traceback is incomplete?!
Would be good to create new issue(s), since this one should have been fixed by 951a9c4.

**Comment by Shougo** [comment]
The original issue is fixed.
Please create the new issue.

@TheSovietStorm Please upload the correct traceback.
